assets/src/js/ScrollSpy.js

```diff
@@ -0,0 +1,193 @@
+
+    static getOffsetTop (element) {
+        let yPosition = 0;
+        let nextElement = element;
+
+        while (nextElement) {
+            yPosition += nextElement.offsetTop;
+            nextElement = nextElement.offsetParent;
+        }
+
+        return yPosition;
+    }
+
+    static getHash (item) {
+        const { hash } = $.first('.c-menu-link', item);
+        return hash;
+    }
+
+    getItemInsideWindow () {
+        let currItem;
+
+        this.items.forEach(item => {
+            const hash = ScrollSpy.getHash(item);
+            const target = document.getElementById(hash.substr(1));
+
+            if (!target) return;
+
+            const isScreenPastSection =
+                window.pageYOffset >= ScrollSpy.getOffsetTop(target) - this.offset;
+            const isScreenBeforeSectionEnd =
+                window.pageYOffset <
+                ScrollSpy.getOffsetTop(target) - this.offset + target.offsetHeight;
+
+            if (this.exact && isScreenPastSection && isScreenBeforeSectionEnd) {
+                currItem = item;
+            }
+            if (!this.exact && isScreenPastSection) {
+                currItem = item;
+            }
+        });
+
+        return currItem;
+    }
+
+    removeActiveClass () {
+        this.items.forEach(item => {
+            item.classList.remove(this.activeClass);
+        });
+    }
+
+    addActiveClass () {
+        if (this.currentItem) {
+            this.currentItem.classList.add(this.activeClass);
+        } else {
+            this.items[0].classList.add(this.activeClass);
+        }
+    }
+
+    onScroll () {
+        this.currentItem = this.getItemInsideWindow();
+
+        if (this.currentItem !== this.lastActiveItem) {
+            this.removeActiveClass();
+            this.lastActiveItem = this.currentItem;
+        }
+
+        this.addActiveClass();
+    }
+
+    scrollTo (target) {
+        return new Promise(resolve => {
+            const targetDistanceFromTop = ScrollSpy.getOffsetTop(target);
+            const startingY = window.pageYOffset;
+            const difference = targetDistanceFromTop - startingY;
+            const easing = this.bezierEasing(...this.bezierEasingValue.split(','));
+            let start = null;
+
+            const step = timestamp => {
+                if (!start) start = timestamp;
+                let progress = timestamp - start;
+                let progressPercentage = progress / this.duration;
+                if (progress >= this.duration) progress = this.duration;
+                if (progressPercentage >= 1) progressPercentage = 1;
+                const perTick = startingY + (easing(progressPercentage) * (difference - this.offset));
+
+                window.scrollTo(0, perTick);
+
+                if (progress < this.duration) {
+                    this.scrollAnimationFrame = window.requestAnimationFrame(step);
+                } else {
+                    window.addEventListener('scroll', this.onScrollHandler);
```

Comment:
Does this need throttling?

Reply:
Done, I applied the throttling up in the constructor. Has to be there so that the event can be removed correctly 😄